alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,391 @@
+        let mut buffer = self.used_buffers.pop().unwrap_or_default();
+
+        buffer.resize(self.payload_offset, 0);
+
+        let encoded_payload_size =
+            bincode::serde::encode_into_std_write(header, &mut buffer, config::standard())?;
+
+        let raw_payload_offset = self.payload_offset + encoded_payload_size;
+
+        buffer.resize(raw_payload_offset + raw_payload_len, 0);
+
+        Ok(Buffer {
+            inner: buffer,
+            raw_payload_offset,
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn send_header_with_payload(&mut self, header: &H, raw_payload: &[u8]) -> Result<()> {
+        let mut buffer = self.get_buffer(header, raw_payload.len())?;
+        buffer.copy_from_slice(raw_payload);
+        self.send(buffer)
+    }
+
+    pub fn send_header(&mut self, header: &H) -> Result<()> {
+        self.send_header_with_payload(header, &[])
+    }
+}
+
+pub struct ReceiverData<H> {
+    buffer: Vec<u8>,
+    payload_offset: usize,
+    used_buffer_queue: mpsc::Sender<Vec<u8>>,
+    had_packet_loss: bool,
+    _phantom: PhantomData<H>,
+}
+
+impl<H> ReceiverData<H> {
+    pub fn had_packet_loss(&self) -> bool {
+        self.had_packet_loss
+    }
+}
+
+impl<H: DeserializeOwned> ReceiverData<H> {
+    pub fn get(&self) -> Result<(H, &[u8])> {
+        let payload = &self.buffer[self.payload_offset..];
+
+        let (header, decoded_size) =
+            bincode::serde::decode_from_slice(payload, config::standard())?;
+
+        Ok((header, &payload[decoded_size..]))
+    }
+
+    pub fn get_header(&self) -> Result<H> {
+        Ok(self.get()?.0)
+    }
+}
+
+impl<H> Drop for ReceiverData<H> {
+    fn drop(&mut self) {
+        self.used_buffer_queue
+            .send(mem::take(&mut self.buffer))
+            .ok();
+    }
+}
+
+pub struct StreamReceiver<H> {
+    payload_offset: usize,
+    packet_receiver: mpsc::Receiver<ReconstructedPacket>,
+    used_buffer_queue: mpsc::Sender<Vec<u8>>,
+    last_packet_index: Option<u32>,
+    _phantom: PhantomData<H>,
+}
+
+// Wrapping comparison for packet indices.
+// Problem: packet indices have a finite bit-width and we have to wrap around upon reaching the
+// maximum value. This function provides proper ordering capability when wrapping around. The
+// maximum index distance that two packets can have is u32::MAX / 2 (which is plenty for any
+// reasonable circumstance).
+fn wrapping_cmp(lhs: u32, rhs: u32) -> Ordering {
+    // Note: since we are using u32, if lhs < rhs then the difference will be closer to u32::MAX
+    let diff = lhs.wrapping_sub(rhs);
+    if diff == 0 {
+        Ordering::Equal
+    } else if diff < u32::MAX / 2 {
+        Ordering::Greater
+    } else {
+        // if diff > u32::MAX / 2, it means the sub operation wrapped
+        Ordering::Less
+    }
+}
+
+/// Get next packet reconstructing from shards.
+/// Returns true if a packet has been recontructed and copied into the buffer.
```

Comment:
Comment seems pretty out of date

Reply:
Deleted. There is nothing worth mentioning anymore